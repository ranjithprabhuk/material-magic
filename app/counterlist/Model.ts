

export interface ICounterList {
  allIds: number[];
  items: ICounter[];
}

export interface ICounter {
  id: number;
  value: number;
  name: string;
}

