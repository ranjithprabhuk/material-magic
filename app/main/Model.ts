import {ICounterList} from '../counterlist';
import {RouterState} from 'react-router-redux';

export interface IAppState {
    counterList: ICounterList;
    routing: RouterState;
}
