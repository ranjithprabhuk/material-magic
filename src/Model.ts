import { RouterState } from 'react-router-redux';
import { ISideBarReduxState } from './app/dashboard/SideBar/ISideBar';
import { IThemeConfig } from './theme/IThemeConfig';

export interface IAppState {
  routing: RouterState;
  sideBar: ISideBarReduxState;
  theme: IThemeConfig;
}
