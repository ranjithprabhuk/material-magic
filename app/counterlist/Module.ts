
import { ICounterList, ICounter } from './Model';
import * as _ from 'lodash';

//
// ActionConstants
//
export type CounterListAddItem = 'counterlist/CounterListAddItem';
export const CounterListAddItem: CounterListAddItem = 'counterlist/CounterListAddItem';

export type CounterListRemoveItem = 'counterlist/CounterListRemoveItem';
export const CounterListRemoveItem: CounterListRemoveItem = 'counterlist/CounterListRemoveItem';

export interface ICounterListActionAddItem {
  type: CounterListAddItem;
}
export interface ICounterListActionRemoveItem {
  type: CounterListRemoveItem;
  id: number;
}

export type CounterListAction = ICounterListActionAddItem | ICounterListActionRemoveItem;


export type CounterIncrement = 'counterlist/CounterIncrement';
export const CounterIncrement: CounterIncrement = 'counterlist/CounterIncrement';

export type CounterDecrement = 'counterlist/CounterDecrement';
export const CounterDecrement: CounterDecrement = 'counterlist/CounterDecrement';

export type CounterNameChange = 'counterlist/CounterNameChange';
export const CounterNameChange: CounterNameChange = 'counterlist/CounterNameChange';


export interface ICounterActionIncrement {
  type: CounterIncrement;
  id: number;
}

export interface ICounterActionDecrement {
  type: CounterDecrement;
  id: number;
}

export interface ICounterActionNameChange {
  type: CounterNameChange;
  id: number;
  newName: string;
}

export type CounterAction = ICounterActionIncrement | ICounterActionDecrement | ICounterActionNameChange;


//
// Action creators
//

export function addCounter():ICounterListActionAddItem {
  return {type: CounterListAddItem};
}

export function removeCounter(id:number):ICounterListActionRemoveItem {
  return {type: CounterListRemoveItem, id: id};
}

export function incrementCounter(id:number):ICounterActionIncrement {
  return {type: CounterIncrement, id:id};
}

export function decrementCounter(id:number):ICounterActionDecrement {
  return {type: CounterDecrement, id:id};
}

export function changeCounterName(id:number, newName:string):ICounterActionNameChange {
  return {type: CounterNameChange, id: id, newName: newName };
}


//
// Reducers
//

const initialCounterListState: ICounterList = {
  allIds: [],
  items: [],
};

const initialCounterState: ICounter = {
  id: 0,
  value: 0,
  name: '',
};

function createNewItem( state: ICounter[] ):ICounter {
  const id = state.length > 0 ? Math.max(...state.map(i => i.id)) + 1 : 0;
  const item = { ... initialCounterState, id: id, name: `Counter #${id}` };
  return item;
}

// Nested counter reducers
export function counterReducers(state: ICounter = initialCounterState, action: CounterAction): ICounter {
  switch (action.type) {
    case CounterIncrement:
      return { ...state, value: state.value + 1 };

    case CounterDecrement:
      return { ...state, value: state.value - 1 };

    case CounterNameChange:
      return { ...state, name: action.newName };

    default:
      return state;
  }
}

// Nested counter list items reducers
function counterListItemsReducer(state: ICounter[], action: CounterListAction | CounterAction): ICounter[] {
  switch (action.type) {
    case CounterListAddItem:
      return [ ...state, createNewItem(state) ];
    case CounterListRemoveItem:
      return _.remove(state, (item) => item.id !== action.id);
    case CounterIncrement:
    case CounterDecrement:
    case CounterNameChange:
      return state.map( item => item.id === action.id ? counterReducers(item, action) : item );
    default:
      return state;
  }
}

// main counterlist state reducer to be combined to root
export default function reducer(state: ICounterList = initialCounterListState, action: CounterListAction | CounterAction): ICounterList {
  const items = counterListItemsReducer(state.items, action);
  return {items: items, allIds: items.map( item => item.id) };
}
