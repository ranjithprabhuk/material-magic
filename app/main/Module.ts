
import { Reducer, combineReducers } from 'redux';
import {IAppState} from './Model';
import {counterListReducer} from '../counterlist';
import {routerReducer} from 'react-router-redux';

export const rootReducer: Reducer<IAppState> = combineReducers<IAppState>({
  counterList: counterListReducer,
  routing: routerReducer,
});
