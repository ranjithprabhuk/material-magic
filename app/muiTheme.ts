
import {MuiTheme, getMuiTheme, spacing, colors} from 'material-ui/styles';
import {fade} from 'material-ui/utils/colorManipulator';

const muiTheme:MuiTheme = getMuiTheme({
    spacing: spacing,
    fontFamily: 'Roboto, sans-serif',
    palette: {
        primary1Color: colors.cyan500,
        primary2Color: colors.cyan700,
        primary3Color: colors.grey400,
        accent1Color: colors.pinkA200,
        accent2Color: colors.grey100,
        accent3Color: colors.grey500,
        textColor: colors.darkBlack,
        alternateTextColor: colors.white,
        canvasColor: colors.white,
        borderColor: colors.grey300,
        disabledColor: fade(colors.darkBlack, 0.3),
        pickerHeaderColor: colors.cyan500,
        clockCircleColor: fade(colors.darkBlack, 0.07),
        shadowColor: colors.fullBlack,
    },
});

export default muiTheme;
