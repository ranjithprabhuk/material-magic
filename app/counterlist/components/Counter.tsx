import * as React from 'react';
import MuiFloatingActionButton from 'material-ui/FloatingActionButton';
import MuiIncreaseIcon from 'material-ui/svg-icons/content/add';
import MuiDecreaseIcon from 'material-ui/svg-icons/content/remove';
import MuiGridList from 'material-ui/GridList/GridList';
import MuiGridTile from 'material-ui/GridList/GridTile';
import MuiTextField from 'material-ui/TextField';
import * as _ from 'lodash';

import {ICounter} from '../Model';
import {incrementCounter, decrementCounter, removeCounter, changeCounterName} from '../Module';
import {IAppState} from '../../main';
import {connect} from 'react-redux';

interface ICounterOwnProps {
  id: number;
}
type ICounterStateProps = ICounter;
interface ICounterDispatchProps {
  onIncrement: typeof incrementCounter;
  onDecrement: typeof decrementCounter;
  onRemove: typeof removeCounter;
  onChangeName: typeof changeCounterName;
}

type ICounterProps = ICounterOwnProps & ICounterStateProps & ICounterDispatchProps;
type ICounterReactState = {
  mouseHover: boolean;
  name: string;
};

class Counter extends React.Component<ICounterProps, ICounterReactState> {
  constructor(props:ICounterProps, context:any) {
    super(props, context);
    this.state = {
      mouseHover: false,
      name: this.props.name,
    };
  }

  public componentWillReceiveProps(props:ICounterProps):void {
    this.setState({ name: props.name });
  }

  public render(): React.ReactElement<{}> {
    const valueStyle: React.CSSProperties = {
      verticalAlign: 'middle',
      display: 'block',
      whiteSpace: 'normal',
      margin: 0,
    };

    const iconStyle: React.CSSProperties = {
      ...valueStyle,
      cursor: 'pointer',
    };


    const tileStyle: React.CSSProperties = {
      textAlign: 'center',
    };

    const {id, value}: ICounterStateProps = this.props;
    const onIncrement = this.props.onIncrement.bind(this, id);
    const onDecrement = this.props.onDecrement.bind(this, id);
    const onRemove = this.props.onRemove.bind(this, id);

    return (
      <MuiGridList cellHeight={90} cols={4}>
        <MuiGridTile style={tileStyle} cols={1} rows={1}>
          <MuiFloatingActionButton data-t-id='decrement' onClick={onDecrement}><MuiDecreaseIcon/></MuiFloatingActionButton>
        </MuiGridTile>
        <MuiGridTile style={tileStyle} cols={2} rows={1}>
          <div data-t-id='hover-area' onMouseEnter={this.handleMouseEnter} onMouseLeave={this.handleMouseLeave}>

            <MuiTextField
              style={{width:'100px'}}
              hintText='Name'
              value={this.state.name}
              onChange={this.handleChange}
              onBlur={this.handleBlur.bind(this)}
            />

            <p style={valueStyle}>Value: <span data-t-id='value'>{value}</span></p>
            { this.state.mouseHover ?
              (<i data-t-id='remove' onClick={onRemove} style={iconStyle} className='material-icons'>delete forever</i>) :
              null
            }
          </div>
        </MuiGridTile>
        <MuiGridTile style={tileStyle} cols={1} rows={1}>
          <MuiFloatingActionButton data-t-id='increment'
                                   secondary={true}
                                   onClick={onIncrement}><MuiIncreaseIcon/></MuiFloatingActionButton>
        </MuiGridTile>
      </MuiGridList>
    );
  }

  //
  // Event handlers
  //

  private handleMouseEnter = () => {
    this.setState({ mouseHover: true });
  }

  private handleMouseLeave = () => {
    this.setState({ mouseHover: false });
  }

  private handleChange = (e:React.FormEvent<HTMLInputElement>) => {
    if( e.currentTarget.value && e.currentTarget.value.length < 13) {
      this.setState({name: e.currentTarget.value});
    }
  }

  private handleBlur = (e:React.FormEvent<HTMLInputElement>) => {
    if (this.state.name && this.props.name !== this.state.name) {
      this.props.onChangeName(this.props.id, this.state.name);
    }
  }

}

const mapStateToProps = (state: IAppState, props: ICounterOwnProps): ICounterStateProps => {
  return _.find<ICounter>(state.counterList.items, (item:ICounter) => item.id === props.id)!;
};

const mapDispatchToProps = {
  onIncrement: incrementCounter,
  onDecrement: decrementCounter,
  onRemove: removeCounter,
  onChangeName: changeCounterName,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Counter);
