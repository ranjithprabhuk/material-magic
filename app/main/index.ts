import {IAppState} from './Model';
import App from './components/App';
import {rootReducer} from './Module';

export {
  IAppState,
  App,
  rootReducer,
};
