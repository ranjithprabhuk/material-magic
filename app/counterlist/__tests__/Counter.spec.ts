import * as React from 'react';
import * as PropTypes from 'prop-types';
import Counter from '../components/Counter';
import {createRenderer, ShallowRenderer} from 'react-test-renderer/shallow';


import {IAppState} from '../../main/Model';
import {mount, ReactWrapper} from 'enzyme';
import muiTheme from '../../muiTheme';



import {Store} from 'react-redux';
import {createStore} from 'redux';
import {rootReducer} from '../../main/Module';

const mountWithContext = (node:any):ReactWrapper<any, any> => mount(node, {
  context: { muiTheme: muiTheme },
  childContextTypes: { muiTheme: PropTypes.object },
});

describe('Counter', () => {
  let renderer:ShallowRenderer;
  let store:Store<IAppState>;
  let counterElement:JSX.Element;

  beforeEach(() => {
    store = createStore<IAppState>(
      rootReducer,
      { counterList: { allIds: [1], items: [{id: 1, value: 0, name: 'Name'}] }, routing: {location:null}},
    );
    renderer = createRenderer();
    counterElement = React.createElement(Counter, {id: 1, store: store} as any) ;
  });

  afterEach(() => {
    // empty
  });

  it('should render without crash', () => {
    expect(renderer.render(counterElement)).toMatchSnapshot();
  });

  describe('when', () => {

    let counter:ReactWrapper<any, any>;
    let value:ReactWrapper<any, any>;
    let increment:ReactWrapper<any, any>;
    let decrement:ReactWrapper<any, any>;
    let hoverArea:ReactWrapper<any, any>;
    let getRemove:() => ReactWrapper<any, any>;

    const init = () => {
      counter = mountWithContext(counterElement);
      value = counter.find('[data-t-id="value"]');
      decrement = counter.find('[data-t-id="decrement"]').find('EnhancedButton');
      increment = counter.find('[data-t-id="increment"]').find('EnhancedButton');
      hoverArea = counter.find('[data-t-id="hover-area"]');
      getRemove =  () => counter.find('[data-t-id="remove"]');
    };

    beforeEach(() => {
      init();
    });

    it('clicking increment should increment counter', () => {
      expect(Number(value.text())).toEqual(0);

      increment.simulate('click');
      expect(Number(value.text())).toEqual(1);

      increment.simulate('click');
      expect(Number(value.text())).toEqual(2);

    });

    it('clicking decrement should decrement counter', () => {
      expect(Number(value.text())).toEqual(0);

      decrement.simulate('click');
      expect(Number(value.text())).toEqual(-1);

      decrement.simulate('click');
      expect(Number(value.text())).toEqual(-2);

    });


    it('hovering "hover area" should show remove button', () => {

      expect( getRemove().exists() ).toBeFalsy();
      hoverArea.simulate('mouseenter');
      expect( getRemove().exists() ).toBeTruthy();

    });

  });
});
