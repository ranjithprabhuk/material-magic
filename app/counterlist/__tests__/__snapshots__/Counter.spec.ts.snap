// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render without crash 1`] = `
<Counter
  id={1}
  name="Name"
  onChangeName={[Function]}
  onDecrement={[Function]}
  onIncrement={[Function]}
  onRemove={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": null,
    }
  }
  value={0}
/>
`;
